test/distributed/test_traceable_collectives.py

```diff
@@ -0,0 +1,236 @@
+from torch._dynamo.utils import same
+from torch._dynamo.testing import CompileCounter
+from torch.fx.experimental.proxy_tensor import make_fx
+from torch.testing._internal.common_distributed import (
+    DynamoDistributedSingleProcTestCase,
+    DynamoDistributedMultiProcTestCase,
+    _dynamo_dist_per_rank_init,
+    requires_nccl,
+    skip_if_lt_x_gpu
+)
+from torch._inductor.compile_fx import compile_fx as inductor_compile_fx
+from torch._inductor.utils import has_triton, run_and_get_triton_code
+import torch._dynamo.logging
+
+# LOL if you don't remember to import this, then the op isn't registered and it hits
+# the no-op C++ kernel that i am forced to implement despite not using it
+import torch.distributed._functional_collectives
+
+
+@requires_nccl()
+class TestCollectivesMultiProc(DynamoDistributedMultiProcTestCase):
+    """
+    Run correctness checks in multi-proc runner, mark with minimum # GPUs to run under
+    """
+    def get_world_trs(self):
+        return {
+            "tag": "",
+            "ranks": list(range(self.world_size)),
+            "group_size": self.world_size,
+        }
+
+    @unittest.skipIf(not has_triton(), "Inductor+gpu needs triton and recent GPU arch")
+    @skip_if_lt_x_gpu(2)
+    # TODO: somehow inductor bg compile threads are causing hangs at exit with distributed work dtor
+    @patch.object(torch._inductor.config, "compile_threads", 1)
+    def test_allreduce_inductor(self):
+        """
+        This is matmul/cat/allreduce is a pattern we aim to optimize.
+        """
+
+        def matmul_cat_col(a, b, c, d, e, f, *, tag, ranks, group_size):
+            x = torch.matmul(a, b)
+            y = torch.matmul(c, d)
+            z = torch.cat((x, y))
+            ar = torch.ops.aten.all_reduce(z, "sum", tag, ranks, group_size)
+            g = torch.matmul(e, f)
+            ar = torch.ops.aten.wait_tensor(ar)
+            out = torch.add(ar, g.repeat(2, 1))
+            return (out, )
+
+        def compile(func, example_inputs):
+            graph = make_fx(func)(*example_inputs)
+            return inductor_compile_fx(graph, example_inputs)
+
+        with _dynamo_dist_per_rank_init(self.rank, self.world_size):
+
+            matmul_cat_col = functools.partial(
+                matmul_cat_col,
+                **self.get_world_trs(),
+            )
+            inputs = (torch.ones(4, 4, device="cuda") + self.rank,) * 6
+
+            # non-ideally, i seem to need to enable this at user level in order to construct a torchdispatch subclass
+            # inside py registered collective ops
+            with enable_python_dispatcher():
+                eager_out = matmul_cat_col(*inputs)
+                compiled_matmul_cat_col = compile(matmul_cat_col, inputs)
+                inductor_out = compiled_matmul_cat_col(*inputs)
+                assert same(eager_out, inductor_out, tol=0.001)
+
+
+@requires_nccl()
+class TestCollectivesInductor(DynamoDistributedSingleProcTestCase):
```

Comment:
pretty interested on how `SingleProcTestCase` works for collective, is it doing allreduce on a single rank?

Reply:
it is calling the allreduce op but maybe the nccl kernel is skipped.

these tests are only measuring whether inductor generates the right code to call into dist.* apis.

I assume the apis will work as intended.

Above there is one 'real' integration test that runs multi-proc